cpp/modmesh/universe/rtree.hpp

```diff
@@ -0,0 +1,700 @@
+    {
+        return !(other.min_x > max_x || other.max_x < min_x ||
+                 other.min_y > max_y || other.max_y < min_y ||
+                 other.min_z > max_z || other.max_z < min_z);
+    }
+
+    bool contains(BoundingBox3D const & other) const
+    {
+        return (other.min_x >= min_x && other.max_x <= max_x &&
+                other.min_y >= min_y && other.max_y <= max_y &&
+                other.min_z >= min_z && other.max_z <= max_z);
+    }
+
+    double get_area() const
+    {
+        return (max_x - min_x) * (max_y - min_y) * (max_z - min_z);
+    }
+
+    double get_enlargement(BoundingBox3D const & addition) const
+    {
+        double new_min_x = std::min(min_x, addition.min_x);
+        double new_min_y = std::min(min_y, addition.min_y);
+        double new_min_z = std::min(min_z, addition.min_z);
+        double new_max_x = std::max(max_x, addition.max_x);
+        double new_max_y = std::max(max_y, addition.max_y);
+        double new_max_z = std::max(max_z, addition.max_z);
+        double new_area = (new_max_x - new_min_x) * (new_max_y - new_min_y) * (new_max_z - new_min_z);
+        return new_area - get_area();
+    }
+
+    void expand(BoundingBox3D const & other)
+    {
+        min_x = std::min(min_x, other.min_x);
+        min_y = std::min(min_y, other.min_y);
+        min_z = std::min(min_z, other.min_z);
+        max_x = std::max(max_x, other.max_x);
+        max_y = std::max(max_y, other.max_y);
+        max_z = std::max(max_z, other.max_z);
+    }
+
+    static BoundingBox3D merge(BoundingBox3D const & a, BoundingBox3D const & b)
+    {
+        return BoundingBox3D(
+            std::min(a.min_x, b.min_x),
+            std::min(a.min_y, b.min_y),
+            std::min(a.min_z, b.min_z),
+            std::max(a.max_x, b.max_x),
+            std::max(a.max_y, b.max_y),
+            std::max(a.max_z, b.max_z));
+    }
+}; /* end of struct BoundingBox3D */
+
+/// Value operations traits for R-tree
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated with T
+template <typename T, typename BoundingBoxType>
+struct RTreeValueOps
+{
+    static BoundingBoxType get_bounding_box(T const & item);
+    static BoundingBoxType get_group_bounding_box(std::vector<T> const & items);
+}; /* end of struct RTreeValueOps */
+
+/// R-tree node structure
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+template <typename T, typename BoundingBoxType, typename ValueOpsType>
+struct RTreeNode
+{
+    BoundingBoxType bbox;
+    std::vector<T> items;
+
+    using RTreeNodeType = RTreeNode<T, BoundingBoxType, ValueOpsType>;
+    std::vector<std::unique_ptr<RTreeNodeType>> nodes;
+
+    RTreeNode(BoundingBoxType const & bbox_in)
+        : bbox(bbox_in)
+    {
+    }
+
+    ~RTreeNode() = default;
```

Comment:
I have a habit to add destructor after constructor.  FYI.  I don't mind you use a different convention.